File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseMergedDateTimeExtractor.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

using Microsoft.Recognizers.Text.Matcher;
using Microsoft.Recognizers.Text.Utilities;
using DateObject = System.DateTime;

namespace Microsoft.Recognizers.Text.DateTime
{
    public class BaseMergedDateTimeExtractor : IDateTimeExtractor
    {
        private static readonly Regex NumberOrConnectorRegex = new Regex(@"^[0-9-]+$", RegexOptions.Compiled);

        private readonly IMergedExtractorConfiguration config;

        public BaseMergedDateTimeExtractor(IMergedExtractorConfiguration config)
        {
            this.config = config;
        }

        public static bool HasTokenIndex(string text, Regex regex, out int index, bool inPrefix)
        {
            index = -1;

            // Support cases has two or more specific tokens
            // For example, "show me sales after 2010 and before 2018 or before 2000"
            // When extract "before 2000", we need the second "before" which will be matched in the second Regex match
            RegexOptions regexFlags = inPrefix ? RegexOptions.RightToLeft | RegexOptions.Singleline : RegexOptions.Singleline;
            var match = Regex.Match(text, regex.ToString(), regexFlags);

            if (match.Success)
            {
                var subStr = inPrefix ? text.Substring(match.Index + match.Length) : text.Substring(0, match.Index);
                if (string.IsNullOrEmpty(subStr))
                {
                    index = inPrefix ? match.Index : match.Length;
                    return true;
                }
            }

            return false;
        }

        public bool TryMergeModifierToken(ExtractResult er, Regex tokenRegex, string text, bool potentialAmbiguity = false)
        {
            var beforeStr = text.Substring(0, er.Start ?? 0);
            var afterStr = text.Substring(er.Start + er.Length ?? 0);

            // Avoid adding mod for ambiguity cases, such as "from" in "from ... to ..." should not add mod
            if (potentialAmbiguity && this.config.AmbiguousRangeModifierPrefix != null && this.config.AmbiguousRangeModifierPrefix.IsMatch(beforeStr))
            {
                var matches = this.config.PotentialAmbiguousRangeRegex.Matches(text).Cast<Match>();

                // Weak ambiguous matches are considered only if the extraction is of type range
                if (matches.Any(m => m.Index < er.Start + er.Length && m.Index + m.Length > er.Start && !(m.Groups[Constants.AmbiguousPattern].Success && !er.Type.EndsWith("range"))))
                {
                    return false;
                }
            }

            if (HasTokenIndex(beforeStr.TrimEnd(), tokenRegex, out var tokenIndex, inPrefix: true))
            {
                var modLength = beforeStr.Length - tokenIndex;

                er.Length += modLength;
                er.Start -= modLength;
                er.Text = text.Substring(er.Start ?? 0, er.Length ?? 0);

                er.Metadata = AssignModMetadata(er.Metadata);

                return true;
            }
            else if (this.config.CheckBothBeforeAfter)
            {
                // check also afterStr
                afterStr = text.Substring(er.Start + er.Length ?? 0);
                if (HasTokenIndex(afterStr.TrimStart(), tokenRegex, out tokenIndex, inPrefix: false))
                {
                    var modLength = tokenIndex + afterStr.Length - afterStr.TrimStart().Length;

                    er.Length += modLength;
                    er.Text = text.Substring(er.Start ?? 0, er.Length ?? 0);
                    er.Data = Constants.HAS_MOD;
                    er.Metadata = AssignModMetadata(er.Metadata);

                    return true;
                }
            }

            return false;
        }

        public List<ExtractResult> Extract(string text)
        {
            return Extract(text, DateObject.Now);
        }

        public List<ExtractResult> Extract(string text, DateObject reference)
        {
            var ret = new List<ExtractResult>();

            if (((this.config.Options & DateTimeOptions.FailFast) != 0) && IsFailFastCase(text))
            {
                // @TODO needs better handling of holidays and timezones.
                // AddTo(ret, this.config.HolidayExtractor.Extract(text, reference), text);
                // ret = AddMod(ret, text);

                return ret;
            }

            var originText = text;
            List<MatchResult<string>> superfluousWordMatches = null;

            // Push
            if ((this.config.Options & DateTimeOptions.EnablePreview) != 0)
            {
                text = MatchingUtil.PreProcessTextRemoveSuperfluousWords(text, this.config.SuperfluousWordMatcher, out superfluousWordMatches);
            }

            // The order is important, since there can be conflicts in merging
            AddTo(ret, this.config.DateExtractor.Extract(text, reference), text);
            AddTo(ret, this.config.TimeExtractor.Extract(text, reference), text);
            AddTo(ret, this.config.DatePeriodExtractor.Extract(text, reference), text);
            AddTo(ret, this.config.DurationExtractor.Extract(text, reference), text);
            AddTo(ret, this.config.TimePeriodExtractor.Extract(text, reference), text);
            AddTo(ret, this.config.DateTimePeriodExtractor.Extract(text, reference), text);
            AddTo(ret, this.config.DateTimeExtractor.Extract(text, reference), text);
            AddTo(ret, this.config.SetExtractor.Extract(text, reference), text);
            AddTo(ret, this.config.HolidayExtractor.Extract(text, reference), text);

            if ((this.config.Options & DateTimeOptions.EnablePreview) != 0)
            {
                AddTo(ret, this.config.TimeZoneExtractor.Extract(text, reference), text);
                ret = this.config.TimeZoneExtractor.RemoveAmbiguousTimezone(ret);
            }

            // This should be the last extraction step, as it needs to determine if the previous text contains time or not
            AddTo(ret, NumberEndingRegexMatch(text, ret), text);

            // Modify time entity to an alternative DateTime expression if it follows a DateTime entity
            if ((this.config.Options & DateTimeOptions.ExtendedTypes) != 0)
            {
                ret = this.config.DateTimeAltExtractor.Extract(ret, text, reference);
            }

            ret = FilterUnspecificDatePeriod(ret);

            // Remove common ambiguous cases
            ret = FilterAmbiguity(ret, text);

            ret = AddMod(ret, text);

            // Filtering
            if ((this.config.Options & DateTimeOptions.CalendarMode) != 0)
            {
                ret = CheckCalendarModeFilters(ret, text);
            }

            ret = ret.OrderBy(p => p.Start).ToList();

            // Merge overlapping results
            ret = ExtractResultExtension.MergeAllResults(ret);

            // Pop
            if ((this.config.Options & DateTimeOptions.EnablePreview) != 0)
            {
                ret = MatchingUtil.PostProcessRecoverSuperfluousWords(ret, superfluousWordMatches, originText);
            }

            return ret;
        }

        private Metadata AssignModMetadata(Metadata metadata)
        {
            if (metadata == null)
            {
                metadata = new Metadata { HasMod = true };
            }
            else
            {
                metadata.HasMod = true;
            }

            return metadata;
        }

        private bool IsFailFastCase(string input)
        {
            return (config.FailFastRegex != null) && (!config.FailFastRegex.IsMatch(input));
        }

        private List<ExtractResult> CheckCalendarModeFilters(List<ExtractResult> ers, string text)
        {
            foreach (var er in ers.Reverse<ExtractResult>())
            {
                foreach (var regex in this.config.TermFilterRegexes)
                {
                    var match = regex.Match(er.Text);
                    if (match.Success)
                    {
                        ers.Remove(er);
                    }
                }
            }

            return ers;
        }

        private void AddTo(List<ExtractResult> dst, List<ExtractResult> src, string text)
        {
            foreach (var result in src)
            {
                if ((config.Options & DateTimeOptions.SkipFromToMerge) != 0)
                {
                    if (ShouldSkipFromToMerge(result))
                    {
                        continue;
                    }
                }

                var isFound = false;
                var overlapIndexes = new List<int>();
                var firstIndex = -1;
                for (var i = 0; i < dst.Count; i++)
                {
                    if (dst[i].IsOverlap(result))
                    {
                        isFound = true;
                        if (dst[i].IsCover(result))
                        {
                            if (firstIndex == -1)
                            {
                                firstIndex = i;
                            }

                            overlapIndexes.Add(i);
                        }
                        else
                        {
                            break;
                        }
                    }
                }

                if (!isFound)
                {
                    dst.Add(result);
                }
                else if (overlapIndexes.Any())
                {
                    var tempDst = dst.Where((_, i) => !overlapIndexes.Contains(i)).ToList();

                    // Insert at the first overlap occurence to keep the order
                    tempDst.Insert(firstIndex, result);
                    dst.Clear();
                    dst.AddRange(tempDst);
                }

                dst.Sort((a, b) => (int)(a.Start - b.Start));
            }
        }

        private bool ShouldSkipFromToMerge(ExtractResult er)
        {
            return config.FromToRegex.IsMatch(er.Text);
        }

        private List<ExtractResult> FilterUnspecificDatePeriod(List<ExtractResult> ers)
        {
            ers.RemoveAll(o => this.config.UnspecificDatePeriodRegex.IsMatch(o.Text));
            return ers;
        }

        private List<ExtractResult> FilterAmbiguity(List<ExtractResult> extractResults, string text)
        {
            if (this.config.AmbiguityFiltersDict != null)
            {
                foreach (var regex in this.config.AmbiguityFiltersDict)
                {
                    foreach (var extractResult in extractResults)
                    {
                        if (regex.Key.IsMatch(extractResult.Text))
                        {
                            var matches = regex.Value.Matches(text).Cast<Match>();
                            extractResults = extractResults.Where(er => !matches.Any(m => m.Index < er.Start + er.Length && m.Index + m.Length > er.Start))
                                .ToList();
                        }
                    }
                }
            }

            // @TODO: Refactor to remove this method and use the general ambiguity filter approach
            extractResults = extractResults.Where(er => !(NumberOrConnectorRegex.IsMatch(er.Text) &&
                    (text.Substring(0, (int)er.Start).Trim().EndsWith("-", StringComparison.Ordinal) || text.Substring((int)(er.Start + er.Length)).Trim().StartsWith("-", StringComparison.Ordinal))))
                    .ToList();

            return extractResults;
        }

        // Handle cases like "move 3pm appointment to 4"
        private List<ExtractResult> NumberEndingRegexMatch(string text, IEnumerable<ExtractResult> extractResults)
        {
            var tokens = new List<Token>();

            foreach (var extractResult in extractResults)
            {
                if (extractResult.Type.Equals(Constants.SYS_DATETIME_TIME, StringComparison.Ordinal) ||
                    extractResult.Type.Equals(Constants.SYS_DATETIME_DATETIME, StringComparison.Ordinal))
                {
                    var stringAfter = text.Substring((int)extractResult.Start + (int)extractResult.Length);
                    var match = this.config.NumberEndingPattern.Match(stringAfter);
                    if (match.Success)
                    {
                        var newTime = match.Groups["newTime"];
                        var numRes = this.config.IntegerExtractor.Extract(newTime.ToString());
                        if (numRes.Count == 0)
                        {
                            continue;
                        }

                        var startPosition = (int)extractResult.Start + (int)extractResult.Length + newTime.Index;
                        tokens.Add(new Token(startPosition, startPosition + newTime.Length));
                    }
                }
            }

            return Token.MergeAllTokens(tokens, text, Constants.SYS_DATETIME_TIME);
        }

        private List<ExtractResult> AddMod(List<ExtractResult> ers, string text)
        {
            foreach (var er in ers)
            {
                // AroundRegex is matched non-exclusively before the other relative regexes in order to catch also combined modifiers e.g. "before around 1pm"
                TryMergeModifierToken(er, config.AroundRegex, text);

                // BeforeRegex in Dutch contains the term "voor" which is ambiguous (meaning both "for" and "before")
                var success = TryMergeModifierToken(er, config.BeforeRegex, text, potentialAmbiguity: true);

                if (!success)
                {
                    success = TryMergeModifierToken(er, config.AfterRegex, text);
                }

                if (!success)
                {
                    // SinceRegex in English contains the term "from" which is potentially ambiguous with ranges in the form "from X to Y"
                    success = TryMergeModifierToken(er, config.SinceRegex, text, potentialAmbiguity: true);
                }

                if (!success)
                {
                    success = TryMergeModifierToken(er, config.EqualRegex, text);
                }

                if (er.Type.Equals(Constants.SYS_DATETIME_DATEPERIOD, StringComparison.Ordinal) ||
                    er.Type.Equals(Constants.SYS_DATETIME_DATE, StringComparison.Ordinal) ||
                    er.Type.Equals(Constants.SYS_DATETIME_TIME, StringComparison.Ordinal))
                {
                    // 2012 or after/above, 3 pm or later
                    var afterStr = text.Substring((er.Start ?? 0) + (er.Length ?? 0));

                    if (afterStr.Length > 1)
                    {

                        var match = config.SuffixAfterRegex.MatchBegin(afterStr.TrimStart(), trim: true);

                        if (match.Success && match.Value != ".")
                        {
                            var isFollowedByOtherEntity = true;

                            if (match.Length == afterStr.Trim().Length)
                            {
                                isFollowedByOtherEntity = false;
                            }
                            else
                            {
                                var nextStr = afterStr.Trim().Substring(match.Length).Trim();
                                var nextEr = ers.FirstOrDefault(t => t.Start > er.Start);

                                if (nextEr == null || !nextStr.StartsWith(nextEr.Text, StringComparison.Ordinal))
                                {
                                    isFollowedByOtherEntity = false;
                                }
                            }

                            if (!isFollowedByOtherEntity)
                            {
                                var modLength = match.Length + afterStr.IndexOf(match.Value, StringComparison.Ordinal);
                                er.Length += modLength;
                                er.Text = text.Substring(er.Start ?? 0, er.Length ?? 0);

                                er.Metadata = AssignModMetadata(er.Metadata);
                            }
                        }
                    }
                }
            }

            return ers;
        }
    }
}

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/extractors/BaseMergedDateTimeExtractor.java
package com.microsoft.recognizers.text.datetime.extractors;

import com.microsoft.recognizers.text.ExtractResult;
import com.microsoft.recognizers.text.Metadata;
import com.microsoft.recognizers.text.datetime.Constants;
import com.microsoft.recognizers.text.datetime.DateTimeOptions;
import com.microsoft.recognizers.text.datetime.extractors.config.IMergedExtractorConfiguration;
import com.microsoft.recognizers.text.datetime.extractors.config.ProcessedSuperfluousWords;
import com.microsoft.recognizers.text.datetime.extractors.config.ResultIndex;
import com.microsoft.recognizers.text.datetime.utilities.ConditionalMatch;
import com.microsoft.recognizers.text.datetime.utilities.MatchingUtil;
import com.microsoft.recognizers.text.datetime.utilities.RegexExtension;
import com.microsoft.recognizers.text.datetime.utilities.Token;
import com.microsoft.recognizers.text.matcher.MatchResult;
import com.microsoft.recognizers.text.utilities.Match;
import com.microsoft.recognizers.text.utilities.MatchGroup;
import com.microsoft.recognizers.text.utilities.RegExpUtility;
import com.microsoft.recognizers.text.utilities.StringUtility;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import org.javatuples.Pair;

public class BaseMergedDateTimeExtractor implements IDateTimeExtractor {

    private final IMergedExtractorConfiguration config;

    @Override
    public String getExtractorName() {
        return "";
    }

    public BaseMergedDateTimeExtractor(IMergedExtractorConfiguration config) {
        this.config = config;
    }

    @Override
    public List<ExtractResult> extract(String input, LocalDateTime reference) {

        List<ExtractResult> ret = new ArrayList<>();
        String originInput = input;
        Iterable<MatchResult<String>> superfluousWordMatches = null;

        if (this.config.getOptions().match(DateTimeOptions.EnablePreview)) {
            ProcessedSuperfluousWords processedSuperfluousWords = MatchingUtil.preProcessTextRemoveSuperfluousWords(input, this.config.getSuperfluousWordMatcher());
            input = processedSuperfluousWords.getText();
            superfluousWordMatches = processedSuperfluousWords.getSuperfluousWordMatches();
        }

        // The order is important, since there is a problem in merging
        addTo(ret, this.config.getDateExtractor().extract(input, reference), input);
        addTo(ret, this.config.getTimeExtractor().extract(input, reference), input);
        addTo(ret, this.config.getDatePeriodExtractor().extract(input, reference), input);
        addTo(ret, this.config.getDurationExtractor().extract(input, reference), input);
        addTo(ret, this.config.getTimePeriodExtractor().extract(input, reference), input);
        addTo(ret, this.config.getDateTimePeriodExtractor().extract(input, reference), input);
        addTo(ret, this.config.getDateTimeExtractor().extract(input, reference), input);
        addTo(ret, this.config.getSetExtractor().extract(input, reference), input);
        addTo(ret, this.config.getHolidayExtractor().extract(input, reference), input);

        if (this.config.getOptions().match(DateTimeOptions.EnablePreview)) {
            addTo(ret, this.config.getTimeZoneExtractor().extract(input, reference), input);
            ret = this.config.getTimeZoneExtractor().removeAmbiguousTimezone(ret);
        }

        // This should be at the end since if need the extractor to determine the previous text contains time or not
        addTo(ret, numberEndingRegexMatch(input, ret), input);

        // modify time entity to an alternative DateTime expression if it follows a DateTime entity
        if (this.config.getOptions().match(DateTimeOptions.ExtendedTypes)) {
            ret = this.config.getDateTimeAltExtractor().extract(ret, input, reference);
        }

        ret = filterUnspecificDatePeriod(ret, input);

        // Remove common ambiguous cases
        ret = filterAmbiguity(ret, input);

        ret = addMod(ret, input);

        // filtering
        if (this.config.getOptions().match(DateTimeOptions.CalendarMode)) {
            checkCalendarFilterList(ret, input);
        }

        ret.sort(Comparator.comparingInt(r -> r.getStart()));

        if (this.config.getOptions().match(DateTimeOptions.EnablePreview)) {
            ret = MatchingUtil.posProcessExtractionRecoverSuperfluousWords(ret, superfluousWordMatches, originInput);
        }

        return ret;
    }

    @Override
    public List<ExtractResult> extract(String input) {
        return this.extract(input, LocalDateTime.now());
    }

    private List<ExtractResult> filterAmbiguity(List<ExtractResult> extractResults, String input) {
        if (config.getAmbiguityFiltersDict() != null) {
            for (Pair<Pattern, Pattern> pair : config.getAmbiguityFiltersDict()) {
                final Pattern key = pair.getValue0();
                final Pattern value = pair.getValue1();

                for (ExtractResult extractResult : extractResults) {
                    Optional<Match> keyMatch = Arrays.stream(RegExpUtility.getMatches(key, extractResult.getText())).findFirst();
                    if (keyMatch.isPresent()) {
                        final Match[] matches = RegExpUtility.getMatches(value, input);
                        extractResults = extractResults.stream()
                            .filter(er -> Arrays.stream(matches).noneMatch(m -> m.index < er.getStart() + er.getLength() && m.index + m.length > er.getStart()))
                            .collect(Collectors.toList());
                    }
                }
            }
        }

        return extractResults;
    }

    private void addTo(List<ExtractResult> dst, List<ExtractResult> src, String text) {
        for (ExtractResult result : src) {
            if (config.getOptions().match(DateTimeOptions.SkipFromToMerge)) {
                if (shouldSkipFromToMerge(result)) {
                    continue;
                }
            }

            boolean isFound = false;
            List<Integer> overlapIndexes = new ArrayList<>();
            int firstIndex = -1;
            for (int i = 0; i < dst.size(); i++) {
                if (dst.get(i).isOverlap(result)) {
                    isFound = true;
                    if (dst.get(i).isCover(result)) {
                        if (firstIndex == -1) {
                            firstIndex = i;
                        }

                        overlapIndexes.add(i);
                    } else {
                        break;
                    }
                }
            }

            if (!isFound) {
                dst.add(result);
            } else if (overlapIndexes.size() > 0) {
                List<ExtractResult> tempDst = new ArrayList<>();
                for (int i = 0; i < dst.size(); i++) {
                    if (!overlapIndexes.contains(i)) {
                        tempDst.add(dst.get(i));
                    }
                }

                // insert at the first overlap occurrence to keep the order
                tempDst.add(firstIndex, result);
                dst.clear();
                dst.addAll(tempDst);
            }

            dst.sort(Comparator.comparingInt(a -> a.getStart()));
        }
    }

    private boolean shouldSkipFromToMerge(ExtractResult er) {
        return Arrays.stream(RegExpUtility.getMatches(config.getFromToRegex(), er.getText())).findFirst().isPresent();
    }

    private List<ExtractResult> numberEndingRegexMatch(String text, List<ExtractResult> extractResults) {
        List<Token> tokens = new ArrayList<>();

        for (ExtractResult extractResult : extractResults) {
            if (extractResult.getType().equals(Constants.SYS_DATETIME_TIME) || extractResult.getType().equals(Constants.SYS_DATETIME_DATETIME)) {
                String stringAfter = text.substring(extractResult.getStart() + extractResult.getLength());
                Pattern numberEndingPattern = this.config.getNumberEndingPattern();
                Optional<Match> match = Arrays.stream(RegExpUtility.getMatches(numberEndingPattern, stringAfter)).findFirst();
                if (match.isPresent()) {
                    MatchGroup newTime = match.get().getGroup("newTime");
                    List<ExtractResult> numRes = this.config.getIntegerExtractor().extract(newTime.value);
                    if (numRes.size() == 0) {
                        continue;
                    }

                    int startPosition = extractResult.getStart() + extractResult.getLength() + newTime.index;
                    tokens.add(new Token(startPosition, startPosition + newTime.length));
                }

            }
        }

        return Token.mergeAllTokens(tokens, text, Constants.SYS_DATETIME_TIME);
    }

    private List<ExtractResult> filterUnspecificDatePeriod(List<ExtractResult> ers, String text) {
        ers.removeIf(er -> Arrays.stream(RegExpUtility.getMatches(config.getUnspecificDatePeriodRegex(), er.getText())).findFirst().isPresent());
        return ers;
    }

    private List<ExtractResult> addMod(List<ExtractResult> ers, String text) {
        int index = 0;

        for (ExtractResult er : ers.toArray(new ExtractResult[0])) {
            MergeModifierResult modifiedToken = tryMergeModifierToken(er, config.getBeforeRegex(), text);

            if (!modifiedToken.result) {
                modifiedToken = tryMergeModifierToken(er, config.getAfterRegex(), text);
            }

            if (!modifiedToken.result) {
                // SinceRegex in English contains the term "from" which is potentially ambiguous with ranges in the form "from X to Y"
                modifiedToken = tryMergeModifierToken(er, config.getSinceRegex(), text, true);
            }

            if (!modifiedToken.result) {
                modifiedToken = tryMergeModifierToken(er, config.getAroundRegex(), text);
            }

            ers.set(index, modifiedToken.er);

            final ExtractResult newEr = modifiedToken.er;

            if (newEr.getType().equals(Constants.SYS_DATETIME_DATEPERIOD) || newEr.getType().equals(Constants.SYS_DATETIME_DATE) || 
                newEr.getType().equals(Constants.SYS_DATETIME_TIME)) {
                
                // 2012 or after/above, 3 pm or later
                String afterStr = text.substring(newEr.getStart() + newEr.getLength()).toLowerCase();

                ConditionalMatch match = RegexExtension.matchBegin(config.getSuffixAfterRegex(), StringUtility.trimStart(afterStr), true);

                if (match.getSuccess()) {
                    boolean isFollowedByOtherEntity = true;

                    if (match.getMatch().get().length == afterStr.trim().length()) {
                        isFollowedByOtherEntity = false;

                    } else {
                        String nextStr = afterStr.trim().substring(match.getMatch().get().length).trim();
                        ExtractResult nextEr = ers.stream().filter(t -> t.getStart() > newEr.getStart()).findFirst().orElse(null);

                        if (nextEr == null || !nextStr.startsWith(nextEr.getText())) {
                            isFollowedByOtherEntity = false;
                        }
                    }

                    if (!isFollowedByOtherEntity) {
                        int modLength = match.getMatch().get().length + afterStr.indexOf(match.getMatch().get().value);
                        int length = newEr.getLength() + modLength;
                        String newText = text.substring(newEr.getStart(), newEr.getStart() + length);

                        er.setMetadata(assignModMetadata(er.getMetadata()));

                        ers.set(index, new ExtractResult(er.getStart(), length, newText, er.getType(), er.getData(), er.getMetadata()));
                    }
                }
            }

            index++;
        }

        return ers;
    }

    private MergeModifierResult tryMergeModifierToken(ExtractResult er, Pattern tokenRegex, String text) {
        return tryMergeModifierToken(er, tokenRegex, text, false);
    }

    private MergeModifierResult tryMergeModifierToken(ExtractResult er, Pattern tokenRegex, String text, boolean potentialAmbiguity) {
        String beforeStr = text.substring(0, er.getStart()).toLowerCase();

        // Avoid adding mod for ambiguity cases, such as "from" in "from ... to ..." should not add mod
        if (potentialAmbiguity &&  config.getAmbiguousRangeModifierPrefix() != null &&
            Arrays.stream(RegExpUtility.getMatches(config.getAmbiguousRangeModifierPrefix(), text)).findFirst().isPresent()) {
            final Match[] matches = RegExpUtility.getMatches(config.getPotentialAmbiguousRangeRegex(), text);
            if (Arrays.stream(matches).anyMatch(m -> m.index < er.getStart() + er.getLength() && m.index + m.length > er.getStart())) {
                return new MergeModifierResult(false, er);
            }
        }

        ResultIndex result = hasTokenIndex(StringUtility.trimEnd(beforeStr), tokenRegex);
        if (result.getResult()) {
            int modLength = beforeStr.length() - result.getIndex();
            int length = er.getLength() + modLength;
            int start = er.getStart() - modLength;
            String newText = text.substring(start, start + length);

            er.setText(newText);
            er.setLength(length);
            er.setStart(start);

            er.setMetadata(assignModMetadata(er.getMetadata()));

            return new MergeModifierResult(true, er);
        }

        return new MergeModifierResult(false, er);
    }

    private Metadata assignModMetadata(Metadata metadata) {

        if (metadata == null) {
            metadata = new Metadata() {
                {
                    setHasMod(true);
                }
            };
        } else {
            metadata.setHasMod(true);
        }

        return metadata;
    }

    private ResultIndex hasTokenIndex(String text, Pattern pattern) {
        Match[] matches = RegExpUtility.getMatches(pattern, text);

        // Support cases has two or more specific tokens
        // For example, "show me sales after 2010 and before 2018 or before 2000"
        // When extract "before 2000", we need the second "before" which will be matched in the second Regex match
        for (Match match : matches) {
            if (StringUtility.isNullOrWhiteSpace(text.substring(match.index + match.length))) {
                return new ResultIndex(true, match.index);
            }
        }

        return new ResultIndex(false, -1);
    }

    private void checkCalendarFilterList(List<ExtractResult> ers, String text) {
        List<ExtractResult> shallowCopy = new ArrayList<>(ers);
        Collections.reverse(shallowCopy);
        for (ExtractResult er : shallowCopy) {
            for (Pattern negRegex : this.config.getFilterWordRegexList()) {
                Optional<Match> match = Arrays.stream(RegExpUtility.getMatches(negRegex, er.getText())).findFirst();
                if (match.isPresent()) {
                    ers.remove(er);
                }
            }
        }
    }

    private class MergeModifierResult {
        public final boolean result;
        public final ExtractResult er;

        private MergeModifierResult(boolean result, ExtractResult er) {
            this.result = result;
            this.er = er;
        }
    }
}
